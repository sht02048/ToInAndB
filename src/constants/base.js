const BASE_PATH = "/assets/sprites/backgrounds/base.png";
const BASE_MODIFIER = {
  GAP: -3,
};

export { BASE_PATH, BASE_MODIFIER };
