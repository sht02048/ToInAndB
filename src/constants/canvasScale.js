const CANVAS_WIDTH = window.innerWidth * 0.6;
const CANVAS_HEIGHT = window.innerHeight;

export { CANVAS_WIDTH, CANVAS_HEIGHT };
