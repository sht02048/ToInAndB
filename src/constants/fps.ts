const FPS = 120;

export default FPS;
