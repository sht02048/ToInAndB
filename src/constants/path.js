const PLAYER_AIRCRAFT_PATH = {
  STATIC: "/assets/sprites/aircraft/player_static.png",
  STRAIGHT: "/assets/sprites/aircraft/player_straight.png",
  LEFT: "/assets/sprites/aircraft/player_left.png",
  RIGHT: "/assets/sprites/aircraft/player_right.png",
};

const BASE_PATH = "/assets/sprites/backgrounds/base.png";

const BLOCK_PATH = {
  LEFT: "/assets/sprites/backgrounds/block_left.png",
  RIGHT: "/assets/sprites/backgrounds/block_right.png",
};

const PLATE_PATH = "/assets/sprites/backgrounds/plate.png";

const SPACE_PATH = "/assets/sprites/backgrounds/space.png";

export { PLAYER_AIRCRAFT_PATH, BASE_PATH, BLOCK_PATH, PLATE_PATH, SPACE_PATH };
