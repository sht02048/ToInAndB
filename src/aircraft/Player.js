import { SPRITE_PATH } from "../constants/path";
import Sprite from "../game/Sprite";

class Player {
  #speed = 3.5;

  constructor(game) {
    this.game = game;

    this.leftShip = new Sprite(SPRITE_PATH.PLATER_LEFT);
    this.rightShip = new Sprite(SPRITE_PATH.PLATER_RIGHT);
    this.staticShip = new Sprite(SPRITE_PATH.PLAYER_STATIC);
    this.straightShip = new Sprite(SPRITE_PATH.PLAYER_STRAIGHT);

    this.currentDirection = this.staticShip;
    this.initialY = this.game.mainCanvas.height;

    this.currentDirection.onload = () => {
      this.x = this.game.mainCanvas.width / 2 - this.currentDirection.width / 2;
      this.y = this.game.mainCanvas.height - this.currentDirection.height * 3;
    };

    this.keyPresses = {
      ArrowUp: false,
      ArrowDown: false,
      ArrowLeft: false,
      ArrowRight: false,
    };

    const handleControl = (event, isDown) => {
      if (Object.hasOwn(this.keyPresses, event.key)) {
        this.keyPresses[event.key] = isDown;
      }
    };

    addEventListener("keydown", (event) => handleControl(event, true));
    addEventListener("keyup", (event) => handleControl(event, false));
  }

  control() {
    this.currentDirection = this.staticShip;

    if (this.keyPresses.ArrowUp) {
      this.currentDirection = this.straightShip;

      if (this.y > this.game.minY) {
        this.y -= this.#speed;
      }
    }

    if (this.keyPresses.ArrowDown && this.y < this.game.maxY) {
      this.y += this.#speed;
    }

    if (this.keyPresses.ArrowLeft) {
      this.currentDirection = this.leftShip;

      if (this.x > this.game.minX) {
        this.x -= this.#speed;
      }
    }

    if (this.keyPresses.ArrowRight) {
      this.currentDirection = this.rightShip;

      if (this.x < this.game.maxX) {
        this.x += this.#speed;
      }
    }
  }

  in() {
    if (this.initialY > 0) {
      this.initialY -= this.game.inAndOutSpeed;
    }
  }

  render() {
    this.game.mainCtx.drawImage(
      this.currentDirection,
      this.x,
      this.y - this.initialY,
    );
  }
}

export default Player;
