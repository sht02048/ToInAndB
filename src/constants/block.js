import { BLOCKS } from "./path";

const LEFT_BLOCKS = [BLOCKS.FIRST_LEFT, BLOCKS.SECOND_LEFT, BLOCKS.THIRD_LEFT];

const RIGHT_BLOCKS = [
  BLOCKS.FIRST_RIGHT,
  BLOCKS.SECOND_RIGHT,
  BLOCKS.THIRD_RIGHT,
];

const BLOCK_LIST = [
  {
    GROUP: "left",
    ITEMS: [
      { img: new Image(), src: BLOCKS.FIRST_LEFT },
      { img: new Image(), src: BLOCKS.SECOND_LEFT },
      { img: new Image(), src: BLOCKS.THIRD_LEFT },
    ],
  },
  {
    GROUP: "right",
    ITEMS: [
      { img: new Image(), src: BLOCKS.FIRST_RIGHT },
      { img: new Image(), src: BLOCKS.SECOND_RIGHT },
      { img: new Image(), src: BLOCKS.THIRD_RIGHT },
    ],
  },
];

const BLOCK_SCALE = 2;
const BLOCK_ID = "block-canvas";

export { BLOCK_LIST, BLOCK_SCALE, BLOCK_ID, LEFT_BLOCKS, RIGHT_BLOCKS };
