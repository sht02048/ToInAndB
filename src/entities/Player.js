import Cockpit from "./Cockpit";
import SpaceShip from "./Spaceship";

import Renderer from "../graphics/Renderer";
import Explosion from "../graphics/Explosion";
import MissileLauncher from "../weapons/MissileLauncher";
import CollisionDetector from "../physics/CollisionDetector";

import TEAM from "../constants/team";
import { PLATER, PROJECTILE } from "../constants/path";
import MISSILE_ROUTE_COMMAND from "../constants/missileRouteCommand";

class Player extends SpaceShip {
  #staticWidth = 35;
  #staticHeight = 49;
  #straightMissileWidth = 36;
  #straightMissileSpeed = 5;
  #straightMissileReload = 15;
  #guidedMissileGap = 80;
  #guidedMissileSpeed = 3;
  #guidedMissileReload = 100;
  #missileDamage = 2;
  #canControl = false;

  constructor() {
    super();

    this.explosion = new Explosion();
    this.leftShip = new Renderer(PLATER.LEFT);
    this.rightShip = new Renderer(PLATER.RIGHT);
    this.staticShip = new Renderer(PLATER.STATIC);
    this.straightShip = new Renderer(PLATER.STRAIGHT);
    this.straightMissileLauncher = new MissileLauncher(
      this.#staticWidth,
      this.#staticHeight,
    );
    this.guidedMissileLauncher = new MissileLauncher(
      this.#staticWidth,
      this.#staticHeight,
    );
    this.straightCollisionDetector = new CollisionDetector(
      this.straightMissileLauncher.missileList,
    );
    this.guidedCollisionDetector = new CollisionDetector(
      this.guidedMissileLauncher.missileList,
    );

    this.currentDirection = this.staticShip;
    this.canvasWidth = this.currentDirection.canvasWidth;
    this.canvasHeight = this.currentDirection.canvasHeight;
    this.spawnX = this.canvasWidth / 2 - this.#staticWidth / 2;
    this.spawnY = this.canvasHeight + this.#staticHeight;

    // ACTIVATE 배포 및 플로우 점검시 주석해제 후 하단에 있는 initialY 삭제 필요
    this.initialY = this.canvasHeight;
    // this.initialY = 0;
    this.x = this.spawnX;
    this.y = this.canvasHeight - this.#staticHeight * 3;

    this.level = 1;
    this.healthPoint = 3;
    this.shipSpeed = 2;
    this.isShooting = false;
    this.reloadFrame = 10;
    this.invincibleFrame = 200;
    this.isInvincible = false;
    this.straightProjectile = PROJECTILE.LEVEL_1;
    this.guidedProjectile = PROJECTILE.GUIDED;

    this.cockpit = new Cockpit(this);
  }

  update() {
    this.frame += 1;

    this.cockpit.checkPlayerStatus(this.level, this.shipSpeed);
    this.straightMissileLauncher.setMissileRoute(
      MISSILE_ROUTE_COMMAND.PLATER_STRAIGHT,
      this.#straightMissileSpeed,
    );
    this.guidedMissileLauncher.setMissileRoute(
      MISSILE_ROUTE_COMMAND.GUIDED,
      this.#guidedMissileSpeed,
    );

    if (!this.#canControl) {
      this.in();
      return;
    }

    if (this.isHit) {
      this.isInvincible = true;
      this.updateExplosion();
      return;
    }

    if (this.isInvincible) {
      this.updateSpawn();
      this.invincibleFrame -= 1;

      if (this.invincibleFrame > 100) {
        return;
      }
    }

    this.launchMissile();
    this.setSize();
    this.upgrade();
    this.cockpit.makeShotSound();
    this.cockpit.control(this.shipSpeed);
    this.straightCollisionDetector.detectCollision();
    this.guidedCollisionDetector.detectCollision();
  }

  render() {
    this.guidedMissileLauncher.render();
    this.straightMissileLauncher.render();

    if (this.isHit) {
      this.explosion.destroy(this.x, this.y, this.#staticWidth);
      return;
    }

    if (this.isInvincible) {
      this.renderSpawn();
      return;
    }

    this.currentDirection.render(this.x, this.y - this.initialY);
  }

  updateExplosion() {
    this.isHit = !this.explosion.isExploded();

    if (!this.isHit) {
      this.x = this.spawnX;
      this.y = this.spawnY;
    }
  }

  updateSpawn() {
    this.#canControl = false;
  }

  renderSpawn() {
    this.currentDirection.mainCtx.save();
    this.currentDirection.mainCtx.globalAlpha = 0.4;
    this.currentDirection.render(this.x, this.y);
    this.currentDirection.mainCtx.restore();
  }

  updateSpawn() {
    if (this.invincibleFrame > 100) {
      this.y -= 2;
    }

    if (this.invincibleFrame < 0) {
      this.isInvincible = false;
      this.invincibleFrame = 200;
    }
  }

  upgrade() {
    switch (this.level) {
      case 2:
        this.#missileDamage = 3;
        this.straightProjectile = PROJECTILE.LEVEL_2;
        this.#straightMissileWidth = 52;
        break;

      case 3:
        this.#missileDamage = 4;
        this.straightProjectile = PROJECTILE.LEVEL_3;
        this.#straightMissileWidth = 90;
        break;
    }

    if (this.reloadFrame % this.#guidedMissileReload === 0 && this.level > 1) {
      this.loadGuidedMissile();
    }
  }

  launchMissile() {
    if (!this.isShooting) {
      return;
    }

    this.reloadFrame += 1;

    if (this.reloadFrame % this.#straightMissileReload !== 0) {
      return;
    }

    this.loadStraightMissile();
  }

  loadStraightMissile() {
    const missileInformation = this.setMissileInformation(
      this.straightProjectile,
      this.x,
      this.y,
      this.#straightMissileWidth,
      this.#straightMissileSpeed,
    );

    this.straightMissileLauncher.loadSingleAmmo(missileInformation);
  }

  loadGuidedMissile() {
    const leftX = this.x - this.currentDirection.width / 2;
    const rightX = this.x + this.currentDirection.width / 2;
    const sharedY = this.y + this.currentDirection.height / 2;

    const leftWidth = this.#guidedMissileGap;
    const rightWidth = -this.#guidedMissileGap / 2;

    const leftMissileInformation = this.setMissileInformation(
      this.guidedProjectile,
      leftX,
      sharedY,
      leftWidth,
      this.#guidedMissileSpeed,
    );
    const rightMissileInformation = this.setMissileInformation(
      this.guidedProjectile,
      rightX,
      sharedY,
      rightWidth,
      this.#guidedMissileSpeed,
    );

    this.guidedMissileLauncher.loadSingleAmmo(leftMissileInformation);
    this.guidedMissileLauncher.loadSingleAmmo(rightMissileInformation);
  }

  in() {
    if (this.initialY < 0) {
      this.#canControl = true;
    }

    this.initialY -= this.currentDirection.inAndOutSpeed;
  }

  setTargetList(targetList) {
    this.straightCollisionDetector.setTargetList(targetList);
    this.guidedCollisionDetector.setTargetList(targetList);
    this.straightMissileLauncher.setTargetList(targetList);
    this.guidedMissileLauncher.setTargetList(targetList);
  }

  setSize() {
    this.width = this.currentDirection.width;
    this.height = this.currentDirection.height;
  }

  setMissileInformation(
    projectilePath,
    positionX,
    positionY,
    missileWidth,
    missileSpeed,
  ) {
    const missileInformation = {
      projectilePath: projectilePath,
      x: positionX,
      y: positionY,
      missileWidth: missileWidth,
      missileDamage: this.#missileDamage,
      missileSpeed: missileSpeed,
      team: TEAM.PLATER,
    };

    return missileInformation;
  }
}

export default Player;
