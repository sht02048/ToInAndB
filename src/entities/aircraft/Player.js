import Projectile from "./Projectile";

import Sprite from "../Sprite";
import { SPRITE_PATH, PROJECTILE_PATH } from "../../constants/path";

class Player {
  #speed = 3.5;

  constructor(game) {
    this.game = game;

    this.leftShip = new Sprite(SPRITE_PATH.PLATER_LEFT);
    this.rightShip = new Sprite(SPRITE_PATH.PLATER_RIGHT);
    this.staticShip = new Sprite(SPRITE_PATH.PLAYER_STATIC);
    this.straightShip = new Sprite(SPRITE_PATH.PLAYER_STRAIGHT);
    this.level = "LEVEL_1";

    this.currentDirection = this.staticShip;

    // ACTIVATE 실제 작업시 주석해제 필요
    // this.initialY = this.game.mainCanvas.height;
    this.initialY = 0;

    this.currentDirection.onload = () => {
      this.x = this.game.mainCanvas.width / 2 - this.currentDirection.width / 2;
      this.y = this.game.mainCanvas.height - this.currentDirection.height * 3;
    };

    this.keyPresses = {
      ArrowUp: false,
      ArrowDown: false,
      ArrowLeft: false,
      ArrowRight: false,
    };
  }

  control() {
    this.currentDirection = this.staticShip;

    if (this.keyPresses.ArrowUp) {
      this.currentDirection = this.straightShip;

      if (this.y > this.game.minY) {
        this.y -= this.#speed;
      }
    }

    if (this.keyPresses.ArrowDown && this.y < this.game.maxY) {
      this.y += this.#speed;
    }

    if (this.keyPresses.ArrowLeft) {
      this.currentDirection = this.leftShip;

      if (this.x > this.game.minX) {
        this.x -= this.#speed;
      }
    }

    if (this.keyPresses.ArrowRight) {
      this.currentDirection = this.rightShip;

      if (this.x < this.game.maxX) {
        this.x += this.#speed;
      }
    }
  }

  in() {
    if (this.initialY > 0) {
      this.initialY -= this.game.inAndOutSpeed;
    }
  }

  addEvent() {
    const handleControl = (event, isDown) => {
      if (Object.hasOwn(this.keyPresses, event.key)) {
        this.keyPresses[event.key] = isDown;
      }
    };

    const handleAttack = (event) => {
      if (event.key === " ") {
        console.log(event);
        this.bullet = new Projectile(PROJECTILE_PATH[this.level]);
        this.bullet.x =
          this.x +
          (this.currentDirection.width * 1.2) / 2 -
          this.bullet.width / 2;
        this.bullet.y = this.y - 50;

        this.game.bulletList.push(this.bullet);
      }
    };

    addEventListener("keydown", (event) => handleControl(event, true));
    addEventListener("keyup", (event) => handleControl(event, false));
    addEventListener("keypress", handleAttack);
  }

  attack() {
    this.game.bulletList.forEach((bullet) => {
      if (bullet.isHitByEnemy) {
        return;
      }

      if (bullet.y > this.game.minY - 50) {
        bullet.y -= bullet.speed;
      }
    });
  }

  render() {
    this.game.mainCtx.drawImage(
      this.currentDirection,
      this.x,
      this.y - this.initialY,
      this.currentDirection.width * 1.2,
      this.currentDirection.height * 1.2,
    );

    this.game.bulletList.forEach((bullet) => {
      if (bullet.isHitByEnemy) {
        return;
      }

      this.game.mainCtx.drawImage(bullet.projectile, bullet.x, bullet.y);
    });
  }
}

export default Player;
