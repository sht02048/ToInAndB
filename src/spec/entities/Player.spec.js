import Player from "../../entities/Player";
import { PROJECTILE } from "../../constants/path";

describe("player", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let player;

  beforeEach(() => {
    player = new Player();
  });

  it("should have proper functions", () => {
    expect(player.update).to.be.a("function");
    expect(player.render).to.be.a("function");
  });

  it("should be rendered on canvas", () => {
    player.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should be destroyed when hit", () => {
    player.isHit = true;
    player.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should be rendered invincible mode", () => {
    player.isInvincible = true;
    player.render();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should be rendered spawn", () => {
    player.renderSpawn();

    const ctx = mainCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  it("should update spawn", () => {
    const previousY = player.y;

    player.updateSpawn();

    const afterY = player.y;

    expect(previousY).to.equal(afterY + 2);

    player.invincibleFrame = -1;
    player.updateSpawn();

    expect(player.invincibleFrame).to.equal(200);
  });

  it("should handle level change", () => {
    player.level = 2;
    player.upgrade();

    expect(player.straightProjectile).to.equal(PROJECTILE.LEVEL_2);

    player.level = 3;
    player.upgrade();

    expect(player.straightProjectile).to.equal(PROJECTILE.LEVEL_3);
  });

  describe("should update", () => {
    it("its status", () => {
      player.healthPoint = 0;
      player.update();

      expect(player.isDestroyed).toEqual(true);
    });

    it("execute out", () => {
      const originalY = player.y;

      player.shouldOut = true;
      player.update();

      expect(originalY).to.not.equal(player.y);
    });
  });

  describe("should launch missiles", () => {
    beforeEach(() => {
      player.isShooting = true;
    });

    it("to straight", () => {
      player.reloadFrame = 15;

      player.launchMissile();

      const ctx = mainCanvas.getContext("2d");
      const events = ctx.__getEvents();

      expect(events).toMatchSnapshot();
    });

    it("to enemy", () => {
      player.reloadFrame = 99;
      player.level = 2;

      player.launchMissile();

      const ctx = mainCanvas.getContext("2d");
      const events = ctx.__getEvents();

      expect(events).toMatchSnapshot();
    });
  });
});
