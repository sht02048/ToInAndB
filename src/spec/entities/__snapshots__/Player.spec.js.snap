// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`player > should be destroyed when hit 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_static.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": 3,
      "img": <img
        src="/images/explosion_enemy_1.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`player > should be rendered invincible mode 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_static.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": 3,
      "img": <img
        src="/images/explosion_enemy_1.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "value": 0.4,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_straight.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;

exports[`player > should be rendered on canvas 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_static.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`player > should be rendered spawn 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_static.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": 3,
      "img": <img
        src="/images/explosion_enemy_1.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "value": 0.4,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_straight.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "value": 0.4,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_straight.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;

exports[`player > should launch missiles > to enemy 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_static.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": 3,
      "img": <img
        src="/images/explosion_enemy_1.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "value": 0.4,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_straight.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "value": 0.4,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_straight.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;

exports[`player > should launch missiles > to straight 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_static.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": 3,
      "img": <img
        src="/images/explosion_enemy_1.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "value": 0.4,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_straight.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "save",
  },
  {
    "props": {
      "value": 0.4,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "globalAlpha",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": 132.5,
      "dy": -147,
      "img": <img
        src="/sprites/player/player_straight.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {},
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "restore",
  },
]
`;
